packager/mux/mux.go

```diff
@@ -37,82 +37,134 @@ import (
+func expectSignerQuery(suffix string, req *http.Request, params *map[string]string, errorMsg *string, errorCode *int) {
+	(*params)["signURL"] = suffix
+	if req.URL.RawQuery != "" {
+		(*params)["signURL"] += "?" + req.URL.RawQuery
+	}
+}
+
+// expectCertQuery is a URL Path Suffix Validator specific to cert requests.
+func expectCertQuery(suffix string, req *http.Request, params *map[string]string, errorMsg *string, errorCode *int) {
+	unescaped, err := url.PathUnescape(suffix)
+	if err != nil {
+		*errorMsg, *errorCode = "400 bad request - bad URL encoding", http.StatusBadRequest
+	} else {
+		(*params)["certName"] = unescaped
+	}
 }
 
-// The main entry point. Use the return value for http.Server.Handler.
+// New is the main entry point. Use the return value for http.Server.Handler.
 func New(certCache http.Handler, signer http.Handler, validityMap http.Handler, healthz http.Handler) http.Handler {
-	return &mux{certCache, signer, validityMap, healthz}
+	return &mux{
+		// Note that the order of rules in the matrix matters: the first
+		// matching rule will be applied, so the rule for “/priv/doc/” precedes
+		// the rule for “/priv/doc”.
+		// Also note that the last rule matches any URL.
+		[]routingRule{
+			{"/priv/doc/", expectSignerQuery, signer, "signer"},
+			{"/priv/doc", expectNoSuffix, signer, "signer"},
+			{"/amppkg/cert/", expectCertQuery, certCache, "certCache"},
+			{"/amppkg/validity", expectNoSuffix, validityMap, "validityMap"},
+			{"/healthz", expectNoSuffix, healthz, "healthz"},
+			{"", return404, nil, "handler_not_assigned"},
+		}}
 }
 
 func tryTrimPrefix(s, prefix string) (string, bool) {
 	sLen := len(s)
 	trimmed := strings.TrimPrefix(s, prefix)
-	return trimmed, len(trimmed) != sLen
+	return trimmed, len(prefix)+len(trimmed) == sLen
 }
 
 var allowedMethods = map[string]bool{http.MethodGet: true, http.MethodHead: true}
 
 func (this *mux) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
-	if !allowedMethods[req.Method] {
-		http.Error(resp, "405 method not allowed", http.StatusMethodNotAllowed)
-		return
-	}
-
-	params := map[string]string{}
-	req = WithParams(req, params)
-
-	// Use EscapedPath rather than RequestURI because the latter can take
-	// absolute-form, per https://tools.ietf.org/html/rfc7230#section-5.3.
-	//
-	// Use EscapedPath rather than RawPath to verify that the request URI
-	// is a valid encoding, per
-	// https://wicg.github.io/webpackage/draft-yasskin-http-origin-signed-responses.html#application-signed-exchange
-	// item 3.
 	path := req.URL.EscapedPath()
-	if suffix, ok := tryTrimPrefix(path, "/priv/doc"); ok {
-		if suffix == "" {
-			this.signer.ServeHTTP(resp, req)
-		} else if suffix[0] == '/' {
-			params["signURL"] = suffix[1:]
-			if req.URL.RawQuery != "" {
-				params["signURL"] += "?" + req.URL.RawQuery
-			}
-			this.signer.ServeHTTP(resp, req)
-		} else {
-			http.NotFound(resp, req)
+
+	// Find the first matching routing rule.
+	matchingRule := (*routingRule)(nil)
```

Comment:
Since you're specifying the type anyway, I would write this as:
```
var matchingRule *routingRule
```
instead of as a cast. (Go default-initializes to the zero value, which for pointers is nil.)